### ISSUE TITLE
pgbouncer unavailable while churning through cancellations

### ISSUE THREAD

**Issue by marcocitus** [pgbouncer/pgbouncer#328]
We've been seeing some stability issues with pgbouncer on busy servers that occur when cancelling queries immediately after a transient failure like an out of memory error.

Basically, we see periods of 5-30 minutes where all pgbouncer returns is:
```
ERROR: pgbouncer cannot connect to server
```

This messages comes from [`check_fast_fail`](https://github.com/pgbouncer/pgbouncer/blob/master/src/objects.c#L601), which as far as I understand means we're in a state where a connection attempt failed and existing connections are dead. As part of this function, pgbouncer also calls `launch_new_connection`.

When opening a connection, `launch_new_connection` immediately [sets `pool->last_connect_time`](https://github.com/pgbouncer/pgbouncer/blob/master/src/objects.c#L1181). Given that `pool->last_connect_failed` is still true, this means other connection attempts [won't proceed](https://github.com/pgbouncer/pgbouncer/blob/master/src/objects.c#L1115) and in the verbose log we see a lot of:

```
2018-09-19 12:05:51.668 3832 DEBUG launch_new_connection: last failed, wait
```

At least we have one connection attempt that is going through, right? Yes, but we have pending cancellations. In that case, `handle_connect` chooses to use that connection to [send a cancellation](https://github.com/pgbouncer/pgbouncer/blob/master/src/server.c#L407) and then (per the protocol) immediately disconnects. 

In the log we see a lot of:
```
2018-09-19 12:02:21.814 3832 DEBUG S-0x1f6f430: citus/citus@(bad-af):0 dns_callback: inet4: 172.16.100.66:5432
2018-09-19 12:02:21.814 3832 DEBUG S-0x1f6f430: citus/citus@172.16.100.66:5432 launching new connection to server
2018-09-19 12:02:21.815 3832 DEBUG S-0x1f6f430: citus/citus@172.16.100.66:5432 S: connect ok
2018-09-19 12:02:21.815 3832 DEBUG S-0x1f6f430: citus/citus@172.16.100.66:5432 use it for pending cancel req
```

Even though we successfully connected to postgres, `pool->last_connect_failed` is not reset in [ `disconnect_server`](https://github.com/pgbouncer/pgbouncer/blob/master/src/objects.c#L837) and since `last_connect_time` was reset there won't be another connection attempt from `launch_new_connection` for another `server_login_retry` (default: 15) seconds (!) and then that connection will be used for the next cancellation in the queue, ...

Basically, with default settings, pgbouncer may need several minutes to reach the end of the cancellation queue before it can answer queries again.

What's a good fix for this? Setting `server_login_retry = 1` does not sufficiently address the stability issue for us. We tried setting `pool->last_connect_failed = 0` within this logic. Unfortunately, that does not seem to work well because then `check_fast_fail` returns immediately and we don't quickly launch new connections when they are needed. We also tried a more hacky approach to set `pool->last_connect_time` to 0 to force `launch_new_connection` to continue. Combined with a low `server_login_retry` that seems to reduce the failure window to a few seconds, but it's kind of hacky.

**Comment by curlup** [comment]
Do you think it might be solved in a way that in a such special case (resembling starting pool from scratch) `handle_connect` somehow didn't choose to use newly established and the only one connection for a pending cancel request?

Another approach comes in mind to introduce some sort of a burst mode to launch_new_connection that won't wait for `server_login_retry` to create new connections?

Btw do you have reproducible setup script or something? 

Also, have you tried non zero `min_pool_size`? Do you think that might work in a way?

**Comment by marcocitus** [comment]
> Do you think it might be solved in a way that in a such special case (resembling starting pool from scratch) handle_connect somehow didn't choose to use newly established and the only one connection for a pending cancel request?

I think it could work in principle, but the fact that pgbouncer prioritises cancellations over queries seems saner than the fact that `launch_new_connection` refuses to open new connections even though postgres is responding to cancellations.

I opened a PR to distinguish the "login failed" from the "connect failed" case to allow `launch_new_connection` to proceed if the last cancellation succeeded even if the last login didn't.

> Btw do you have reproducible setup script or something?

I do in my own environment (Citus Cloud), but that's a bit hard to replicate. Citus parallelises queries over many connections and on Citus Cloud those connection go through outbound pgbouncers (one per worker). If a task (query on a shard) fails, Citus first retries the task, and if it fails again Citus cancels all pending tasks. When I run a parallel query that generates an out-of-memory error on the worker several times, it eventually creates the condition where there are several cancellations in the queue while a connection attempt fails (e.g. postgres fails to fork due to out-of-memory). Then the pgbouncer becomes unavailable for a few minutes. 

I'll see if I can write a script that exhibits the same cancellation behaviour.

> Also, have you tried non zero min_pool_size? Do you think that might work in a way?

Yes, but that does not work because the janitor just calls `launch_new_connection` which won't open a new connection as long as the `last_connect_failed` flag is set.

**Comment by svb007** [comment]
How different is this problem from the one we tried to solve in https://github.com/pgbouncer/pgbouncer/issues/285 ?

**Comment by marcocitus** [comment]
> How different is this problem from the one we tried to solve in #285 ?

It seems mostly unrelated since we're not using PAM and the problematic code is in a different place.
